### ISSUE TITLE
if you want to port on Mac OS : Syscall

### ISSUE THREAD

**Issue by smougel** [nemasu/asmttpd#6]
http://www.opensource.apple.com/source/xnu/xnu-1456.1.26/bsd/kern/syscalls.master

**Comment by nickdesaulniers** [comment]
I was hacking on this a little, I'll upstream some of my patches.  The first step was to get everything to build, which I have.

**Comment by nickdesaulniers** [comment]
syscalls used:

| name | [Linux]() | [Apple](http://www.opensource.apple.com/source/xnu/xnu-1456.1.26/bsd/kern/syscalls.master) |
| sigaction | 13 | 46 |
...

**Comment by nickdesaulniers** [comment]
I think we should use the preprocessor to check for [OS specific symbols](http://nadeausoftware.com/articles/2012/01/c_c_tip_how_use_compiler_predefined_macros_detect_operating_system#OSXiOSandDarwin) and have conditional assembling.  [example](https://github.com/girino/jhProtominer-bad/blob/ca99ef58f3235990c457f20133983ec01ed643c1/src/intel/sha256_avx1.asm#L416)

**Comment by nickdesaulniers** [comment]
looks like the system call number need to be shifted.  I have verified this, will try to find more info as to why: https://filippo.io/making-system-calls-from-assembly-in-mac-os-x/

**Comment by nickdesaulniers** [comment]
I'm able to get the usage warning printing! w00t

**Comment by nickdesaulniers** [comment]
does linux pass command line args on the stack?
https://github.com/nemasu/asmttpd/blob/master/main.asm#L53-L58

When I inspect the registers on entry, %rdi is argc and %rsi is argv.  The current code looks like it pulls them off the stack.  Calling print_line with start_text blows away %rdi and %rsi.

**Comment by nemasu** [comment]
argv and argc are passed on the stack.

But for Linux system calls it uses registers first, then stack:

https://github.com/nemasu/asmttpd/blob/master/main.asm#L28

Which is what I use internally too.

**Comment by nickdesaulniers** [comment]
> argv and argc are passed on the stack.

right, but it doesn't look to be the case for Mach-O binaries.  Asking an friend who's an LLVM dev to confirm.

I'll need this for OSX's equivalent to nanosleep:
https://opensource.apple.com/source/Libc/Libc-583/gen/nanosleep.c

**Comment by nickdesaulniers** [comment]
err, they're passed on the stack for 32b programs on Linux/OSX and in %rdi and %rsi for 64b programs.

The makefile makes it looks like you're compiling a 64b binary, so why are argc/argv being popped off the stack?

**Comment by nemasu** [comment]
According to the AMD64 ABi doc, http://www.x86-64.org/documentation/abi.pdf,  page 29, argv and argc are passed on the stack.

**Comment by nickdesaulniers** [comment]
I'd love to trust a piece of paper over what the compiler as actually doing. But if we take a look at the dissasembly on OSX:

```
Disassembly of section .text:

0000000100000f60 <_main>:
#include <stdio.h>
int main (int argc, char** argv) {
   100000f60:	55                   	push   %rbp
   100000f61:	48 89 e5             	mov    %rsp,%rbp
  puts(argv[0]);
   100000f64:	48 8b 3e             	mov    (%rsi),%rdi [hello! taking argv[0] from %rsi and putting in %rdi for puts]
   100000f67:	e8 04 00 00 00       	callq  100000f70 <_main+0x10>
   100000f6c:	31 c0                	xor    %eax,%eax
}
```

**Comment by nickdesaulniers** [comment]
> Unless _start is pulling things off the stack and into registers before calling into _main.

That MUST be the case.  Without the `-osx_min_version` flag to ld, arguments are indeed passed on stack to _start.  Ok, I'll carry on.

**Comment by nickdesaulniers** [comment]
For my reference, `man syscall` and `/usr/include/sys/syscall.h` are helpful!

**Comment by nickdesaulniers** [comment]
Looks like _start is defined in `/usr/lib/crt1.o`, which should be linked against, according to `man 1 ld`.

**Comment by nickdesaulniers** [comment]
I've got a lot more of this working.  The real work will be pthread_create (processes and threads are created with clone(2) on Linux, not so on OSX), and nanosleep (nanosleep(2) on Linux, __semwait_signal) on OSX.  I think I'll need to link against libc for OSX to get the address of clock_sem.

**Comment by nickdesaulniers** [comment]
it looks like the pthread_create+pthread_join combo translates into a combo of syscalls:
* proc_info
* bsdthread_create
* thread_selfid
* __disable_threadsignal
* __semwait_signal

**Comment by nickdesaulniers** [comment]
[proc_info](https://opensource.apple.com/source/xnu/xnu-1228.9.59/bsd/kern/proc_info.c)
[bsdthread_create](https://opensource.apple.com/source/xnu/xnu-1228.9.59/bsd/kern/pthread_synch.c)

**Comment by DmitryHetman** [comment]
gcc on mac works, so you can use as as assembler.

**Comment by DmitryHetman** [comment]
as - Portable GNU assembler.

**Comment by nickdesaulniers** [comment]
What is your point?  Both comments seem unrelated.  See my comment about nanosleep and clone.  Also, I sold off my last piece of Mac hardware, so I will no longer be pursuing this issue. :(
